File: src/rex/states/processing.py
"""
Processing state handler.

Sends the user's query to the agent and handles the response.
"""

from typing import TYPE_CHECKING

from agent import PendingConfirmation, run_voice_agent
from audio.feedback import ThinkingTone
from core.state_machine import ConversationState, StateHandler, StateResult

if TYPE_CHECKING:
    from core.context import AppContext


class ProcessingHandler(StateHandler):
    """
    Handler for the PROCESSING state.

    Invokes the LLM agent with the user's query and determines
    the next state based on the response.
    """

    def __init__(self):
        self._transcription = ""

    @property
    def state(self) -> ConversationState:
        return ConversationState.PROCESSING

    def enter(self, ctx: "AppContext", data: dict | None = None) -> None:
        """Store the transcription to process."""
        if data:
            self._transcription = data.get("transcription", "")
        else:
            self._transcription = ""

    def process(self, ctx: "AppContext") -> StateResult:
        """
        Process the query with the agent.

        Returns:
            SPEAKING with response text
            AWAITING_CONFIRMATION if tool needs confirmation
            WAITING_FOR_WAKE_WORD on error
        """
        if not self._transcription:
            return StateResult(next_state=ConversationState.WAITING_FOR_WAKE_WORD)

        try:
            print("🤔 Thinking...")

            # Play thinking tone while waiting for LLM
            with ThinkingTone():
                result, history, thread_id = run_voice_agent(
                    self._transcription,
                    ctx.conversation_history,
                    ctx.thread_id,
                )

            # Update conversation state
            ctx.conversation_history = history
            ctx.thread_id = thread_id

            # Check if we need confirmation
            if isinstance(result, PendingConfirmation):
                return StateResult(
                    next_state=ConversationState.AWAITING_CONFIRMATION,
                    data={"pending": result},
                )

            # Normal response
            response = result
            return StateResult(
                next_state=ConversationState.SPEAKING,
                data={"response": response},
            )

        except Exception as e:
            print(f"❌ Agent error: {e}")
            import traceback

            traceback.print_exc()

            return StateResult(
                next_state=ConversationState.SPEAKING,
                data={
                    "response": "Sorry, I encountered an error processing your request.",
                    "force_end_conversation": True,
                },
            )

    def exit(self, ctx: "AppContext") -> None:
        """Clear transcription."""
        self._transcription = ""

File: internal/states/processing.go
package states

import (
	"context"
	"log/slog"
)

// ProcessingHandler implements the PROCESSING state.
// It sends the user's transcription to the LLM agent and routes to the
// appropriate next state based on the response.
type ProcessingHandler struct {
	deps          *Deps
	transcription string
}

// NewProcessingHandler creates a ProcessingHandler.
func NewProcessingHandler(deps *Deps) *ProcessingHandler {
	return &ProcessingHandler{deps: deps}
}

func (h *ProcessingHandler) State() int { return StateProcessing }

// Enter stores the transcription provided by the listening state.
func (h *ProcessingHandler) Enter(data map[string]any) {
	h.transcription = ""
	if data != nil {
		if v, ok := data["transcription"].(string); ok {
			h.transcription = v
		}
	}
}

// Process invokes the agent with the user's query and determines the
// next state from the result.
func (h *ProcessingHandler) Process(ctx context.Context) StateResult {
	if h.transcription == "" {
		return StateResult{NextState: StateWaitingForWakeWord}
	}

	slog.Info("processing query", "text", h.transcription)
	h.deps.Audio.StartThinkingTone()

	result, err := h.deps.Agent.Run(ctx, h.transcription, h.deps.History, h.deps.ThreadID)

	h.deps.Audio.StopThinkingTone()

	if err != nil {
		slog.Error("agent error", "err", err)
		return StateResult{
			NextState: StateSpeaking,
			Data: map[string]any{
				"response":               "Sorry, I encountered an error processing your request.",
				"force_end_conversation": true,
			},
		}
	}

	// Update shared conversation state.
	h.deps.History = result.History
	h.deps.ThreadID = result.ThreadID

	if result.PendingConfirmation != nil {
		return StateResult{
			NextState: StateAwaitingConfirmation,
			Data:      map[string]any{"pending": result.PendingConfirmation},
		}
	}

	return StateResult{
		NextState: StateSpeaking,
		Data:      map[string]any{"response": result.Response},
	}
}

// Exit clears the stored transcription.
func (h *ProcessingHandler) Exit() {
	h.transcription = ""
}
